### PR TITLE
schedule: Provide logs when the fifo job panic happens

### PR DESCRIPTION
**Pull request by SimFG**
To make the fifo scheduler better debuggability.

### CODE REVIEW

File: pkg/schedule/schedule.go [hunk 1]
```diff
@@ -149,17 +156,29 @@ func (f *fifo) run() {
...
+		if !ctxDone {
 			f.finishCond.L.Lock()
 			f.finished++
 			f.pendings = f.pendings[1:]
 			f.finishCond.Broadcast()
 			f.finishCond.L.Unlock()
 		}
-	}
+		if err := recover(); err != nil {
+			f.lg.Warn("execute job failed", zap.Any("job", todo), zap.Any("err", err))
```
Comment (serathius) [inline pkg/schedule/schedule.go:179]: What do you expect will be printed for `job` key? Just guessing that it will be pointer to function. In most cases we pass anonymous function to fifo. 

Maybe we could add a proper name to jobs scheduled? Could be a separate PR

Comment (SimFG) [reply pkg/schedule/schedule.go:179]: I want to give some tips when the panic happens. But I don’t have a good idea. Maybe deleting the job key is better and we add a proper name to jobs in a separate pr. What do you think of this?

File: pkg/schedule/schedule.go [hunk 2]
```diff
@@ -56,14 +58,20 @@ type fifo struct {
...
 	finishCond *sync.Cond
 	donec      chan struct{}
+	lg         *zap.Logger
 }
 
 // NewFIFOScheduler returns a Scheduler that schedules jobs in FIFO
 // order sequentially
-func NewFIFOScheduler() Scheduler {
+func NewFIFOScheduler(lg *zap.Logger) Scheduler {
+	if lg == nil {
```
Comment (ptabor) [inline pkg/schedule/schedule.go:67]: We decided to `verify.Assert` (client/pkg/verify/verify.go) whether the logger is prided rather than silently turn off logging if it was not provided.

File: pkg/schedule/schedule.go [hunk 3]
```diff
@@ -149,17 +156,29 @@ func (f *fifo) run() {
...
 				return
 			}
 		} else {
-			todo(f.ctx)
+			f.executeJob(todo, false)
+		}
+	}
+}
+
+func (f *fifo) executeJob(todo Job, ctxDone bool) {
```
Comment (ptabor) [inline pkg/schedule/schedule.go:169]: nit: Instead of `ctxDone` I would call it `updatedFinishedStats` as we should talk about meaning of the parameter, and not why we had 'set' it.

File: pkg/schedule/schedule.go [hunk 4]
```diff
@@ -149,17 +156,29 @@ func (f *fifo) run() {
...
+	defer func() {
+		if !ctxDone {
 			f.finishCond.L.Lock()
 			f.finished++
 			f.pendings = f.pendings[1:]
 			f.finishCond.Broadcast()
 			f.finishCond.L.Unlock()
 		}
-	}
+		if err := recover(); err != nil {
```
Comment (ptabor) [inline pkg/schedule/schedule.go:178]: I have a doubts about the `catching exceptions and swallowing errors` behavior.
If one of the periodic book-keeping processes within DB server fails unexpectedly, it might be caused by assertion that is supposed to detect data inconsistency or other serious problem (e.g. during compaction).   The earlier we detect it and avoid propagation (by e.g. reading checkpoint state from the other replica), the lowest chances we will make this 'broken' state authoritative and the easier to debug the problem. 

How about expanding the logging (as you did), flushing the logs, but applying this pattern: 

https://stackoverflow.com/questions/34842322/can-i-recover-from-panic-handle-the-error-then-panic-again-and-keep-the-origin

instead of the `recover` ?

We should look at this error 1 by 1 and decide whether they are recoverable or not.

**Review by ptabor — CHANGES_REQUESTED** [review]

File: pkg/schedule/schedule.go (hunk 4 above)
Comment (SimFG) [reply pkg/schedule/schedule.go:178]: Got it. About the `recover`, I write a demo.
image

image

So I think the recover method can be used. So I plan to add a `HandlePanic` func in the `Job` struct. When the `HandlePanic` isn't nil, it will be called when the panic happens.

File: pkg/schedule/schedule.go [hunk 5]
```diff
@@ -149,17 +160,34 @@ func (f *fifo) run() {
...
+		if !updatedFinishedStats {
 			f.finishCond.L.Lock()
 			f.finished++
 			f.pendings = f.pendings[1:]
 			f.finishCond.Broadcast()
 			f.finishCond.L.Unlock()
 		}
-	}
+		if err := recover(); err != nil {
+			f.lg.Warn("execute job failed", zap.String("job", todo.Name), zap.Any("err", err))
```
Comment (ptabor) [inline pkg/schedule/schedule.go:183]: I would make it Error... not Warn, as this type of error "used" to panic the server.

File: pkg/schedule/schedule.go [hunk 6]
```diff
@@ -149,17 +160,34 @@ func (f *fifo) run() {
...
+			if todo.HandlePanic != nil {
+				todo.HandlePanic(err)
+			} else {
+				panic(err)
```
Comment (ptabor) [inline pkg/schedule/schedule.go:187]: lg.Panic will trigger flushing logs, so is recommended from debuggability standpoint.

File: pkg/schedule/schedule.go [hunk 7]
```diff
@@ -17,15 +17,23 @@ package schedule
...
 	"context"
 	"sync"
+
+	"go.etcd.io/etcd/client/pkg/v3/verify"
+
+	"go.uber.org/zap"
 )
 
-type Job func(context.Context)
+type Job struct {
```
Comment (ptabor) [inline pkg/schedule/schedule.go:26]: no strong opinion, but probably I would make it an interface... 

and have a:
   `NewJob("foo", func() {...} )` constructor...
   `NewRecoveringJob("boo", func() {...}, func() { ... } );

File: pkg/schedule/schedule.go (hunk 4 above)
Comment (ptabor) [reply pkg/schedule/schedule.go:182]: Thank you. This seems good overall. 

Minor doubt I have... is what's the benefit of having separate HandlePanic(...) if Job author could actually already handle it within the Job.Do itself.

Comment (SimFG) [reply pkg/schedule/schedule.go:182]: Thanks your reading. It seems that the `HandlePanic ` method is a little redundant.

File: pkg/schedule/schedule.go (hunk 7 above)
Comment (SimFG) [reply pkg/schedule/schedule.go]: Thanks for your advice.

File: server/etcdserver/server.go [hunk 8]
```diff
@@ -839,7 +839,7 @@ func (s *EtcdServer) run() {
 	for {
 		select {
 		case ap := <-s.r.apply():
-			f := func(context.Context) { s.applyAll(&ep, &ap) }
+			f := schedule.NewJob("server applyAll", func(context.Context) { s.applyAll(&ep, &ap) })
```
Comment (ptabor) [inline server/etcdserver/server.go:842]: nit: I would avoid using 'space' in the identifiers. 

I would prefer any of:  "server.applyAll" or "server:applyAll" or "server_applyAll" as the convention for naming.   If we ever used this IDs for monitoring metric's field, it would also make things cleaner.

File: server/storage/mvcc/kvstore.go [hunk 9]
```diff
@@ -202,7 +202,7 @@ func (s *store) updateCompactRev(rev int64) (<-chan struct{}, int64, error) {
 	s.revMu.Lock()
 	if rev <= s.compactMainRev {
 		ch := make(chan struct{})
-		f := func(ctx context.Context) { s.compactBarrier(ctx, ch) }
+		f := schedule.NewJob("kvstore updateCompactRev compactBarrier", func(ctx context.Context) { s.compactBarrier(ctx, ch) })
```
Comment (ptabor) [inline server/storage/mvcc/kvstore.go:205]: ditto.

File: server/storage/mvcc/kvstore.go [hunk 10]
```diff
@@ -225,7 +225,7 @@ func (s *store) updateCompactRev(rev int64) (<-chan struct{}, int64, error) {
 
 func (s *store) compact(trace *traceutil.Trace, rev, prevCompactRev int64) (<-chan struct{}, error) {
 	ch := make(chan struct{})
-	var j = func(ctx context.Context) {
+	j := schedule.NewJob("kvstore compact", func(ctx context.Context) {
```
Comment (ptabor) [inline server/storage/mvcc/kvstore.go:228]: ditto

**Review by ptabor — APPROVED** [review]
Thank you. Apart of naming convention, looks good for merging.

File: server/etcdserver/server.go (hunk 8 above)
Comment (SimFG) [reply server/etcdserver/server.go]: I have changed it.

**Review by serathius — APPROVED** [review]

File: pkg/schedule/schedule.go [hunk 11]
```diff
@@ -149,17 +179,29 @@ func (f *fifo) run() {
...
+		if !updatedFinishedStats {
 			f.finishCond.L.Lock()
 			f.finished++
 			f.pendings = f.pendings[1:]
 			f.finishCond.Broadcast()
 			f.finishCond.L.Unlock()
 		}
-	}
+		if err := recover(); err != nil {
+			f.lg.Panic("execute job failed", zap.String("job", todo.Name()), zap.Any("err", err))
```
[the remark this answers is not in the document]
Comment (serathius) [reply pkg/schedule/schedule.go:202]: Have you tested how this log will look like?

[the remark this answers is not in the document]
Comment (SimFG) [reply pkg/schedule/schedule.go:202]: image
 the `err` can't use the zap.Error(), because the `zap.Error()` need a param of the `error` type. In this code, the type of the `err` is `Any`.

[the remark this answers is not in the document]
Comment (serathius) [reply pkg/schedule/schedule.go:202]: ack, then maybe we shouldn't use `err`. Using it could potentially pass different type under, maybe instead use `panic` key?

[the remark this answers is not in the document]
Comment (SimFG) [reply pkg/schedule/schedule.go]: Good idea. I have changed it.
